sql/opt/exec/execbuilder: TestExecBuild failed

[(sql/opt/exec/execbuilder).TestExecBuild failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2597711&tab=buildLog) on [master@3d01fda046c609ee5947e2f41c099dae35b206c1](https://github.com/cockroachdb/cockroach/commits/3d01fda046c609ee5947e2f41c099dae35b206c1):

```
=== RUN   TestExecBuild
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExecBuild524875429
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestExecBuild
    logic.go:3300: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExecBuild524875429
--- FAIL: TestExecBuild (477.53s)
=== RUN   TestExecBuild/local/autocommit_nonmetamorphic
    logic.go:2300: let $rangeid = 35
E210120 06:09:55.348246 6201 sql/logictest/logic.go:3507  

testdata/autocommit_nonmetamorphic:107: SELECT operation, message FROM [SHOW KV TRACE FOR SESSION]
WHERE message     LIKE '%r35: sending batch%'
  AND message NOT LIKE '%PushTxn%'
  AND message NOT LIKE '%QueryTxn%'
expected:
    dist sender send  r35: sending batch 2 CPut, 1 EndTxn to (n1,s1):1
    
but found (query options: "") :
    dist sender send  r35: sending batch 2 CPut to (n1,s1):1
    dist sender send  r35: sending batch 2 CPut, 1 EndTxn to (n1,s1):1
    logic.go:2259: 
         
        testdata/autocommit_nonmetamorphic:107: SELECT operation, message FROM [SHOW KV TRACE FOR SESSION]
        WHERE message     LIKE '%r35: sending batch%'
          AND message NOT LIKE '%PushTxn%'
          AND message NOT LIKE '%QueryTxn%'
        expected:
            dist sender send  r35: sending batch 2 CPut, 1 EndTxn to (n1,s1):1
            
        but found (query options: "") :
            dist sender send  r35: sending batch 2 CPut to (n1,s1):1
            dist sender send  r35: sending batch 2 CPut, 1 EndTxn to (n1,s1):1
--- done: testdata/autocommit_nonmetamorphic with config local: 16 tests, 1 failures
E210120 06:09:55.351044 6201 sql/logictest/logic.go:3532  testdata/autocommit_nonmetamorphic:115: too many errors encountered, skipping the rest of the input
    logic.go:3085: 
        testdata/autocommit_nonmetamorphic:115: error while processing
    logic.go:3085: testdata/autocommit_nonmetamorphic:115: too many errors encountered, skipping the rest of the input
        --- FAIL: TestExecBuild/local/autocommit_nonmetamorphic (1.12s)
=== RUN   TestExecBuild/local
    --- FAIL: TestExecBuild/local (214.13s)

```

<details><summary>More</summary><p>


```
make stressrace TESTS=TestExecBuild PKG=./pkg/sql/opt/exec/execbuilder TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestExecBuild.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/opt/exec/execbuilder).TestExecBuild failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2598340&tab=buildLog) on [master@c6dff23997faeebc8f43df5f8fe8f5393dd38dd1](https://github.com/cockroachdb/cockroach/commits/c6dff23997faeebc8f43df5f8fe8f5393dd38dd1):

```
=== RUN   TestExecBuild
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExecBuild324162892
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestExecBuild
    logic.go:3300: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExecBuild324162892
--- FAIL: TestExecBuild (47.30s)
=== RUN   TestExecBuild/local/autocommit_nonmetamorphic
=== PAUSE TestExecBuild/local/autocommit_nonmetamorphic
=== CONT  TestExecBuild/local/autocommit_nonmetamorphic
=== CONT  TestExecBuild/local/autocommit_nonmetamorphic
    logic.go:2300: let $rangeid = 35
E210120 07:52:26.553625 84 sql/logictest/logic.go:3507  

testdata/autocommit_nonmetamorphic:107: SELECT operation, message FROM [SHOW KV TRACE FOR SESSION]
WHERE message     LIKE '%r35: sending batch%'
  AND message NOT LIKE '%PushTxn%'
  AND message NOT LIKE '%QueryTxn%'
expected:
    dist sender send  r35: sending batch 2 CPut, 1 EndTxn to (n1,s1):1
    
but found (query options: "") :
    dist sender send  r35: sending batch 2 CPut to (n1,s1):1
    dist sender send  r35: sending batch 2 CPut, 1 EndTxn to (n1,s1):1
    logic.go:2259: 
         
        testdata/autocommit_nonmetamorphic:107: SELECT operation, message FROM [SHOW KV TRACE FOR SESSION]
        WHERE message     LIKE '%r35: sending batch%'
          AND message NOT LIKE '%PushTxn%'
          AND message NOT LIKE '%QueryTxn%'
        expected:
            dist sender send  r35: sending batch 2 CPut, 1 EndTxn to (n1,s1):1
            
        but found (query options: "") :
            dist sender send  r35: sending batch 2 CPut to (n1,s1):1
            dist sender send  r35: sending batch 2 CPut, 1 EndTxn to (n1,s1):1
--- done: testdata/autocommit_nonmetamorphic with config local: 16 tests, 1 failures
E210120 07:52:26.553907 84 sql/logictest/logic.go:3532  testdata/autocommit_nonmetamorphic:115: too many errors encountered, skipping the rest of the input
    logic.go:3085: 
        testdata/autocommit_nonmetamorphic:115: error while processing
    logic.go:3085: testdata/autocommit_nonmetamorphic:115: too many errors encountered, skipping the rest of the input
        --- FAIL: TestExecBuild/local/autocommit_nonmetamorphic (0.73s)
=== RUN   TestExecBuild/local
    --- FAIL: TestExecBuild/local (0.00s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestExecBuild PKG=./pkg/sql/opt/exec/execbuilder TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestExecBuild.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/ambiguous_commit_test.go
+++ b/pkg/sql/ambiguous_commit_test.go
@@ -64,7 +64,7 @@ func TestAmbiguousCommit(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	if mutations.MaxBatchSize() == 1 {
+	if mutations.MaxBatchSize(false /* forceProductionMaxBatchSize */) == 1 {
 		// This test relies on the fact that the mutation batch consisting of a
 		// single row also contains an EndTxn which is the case only when the
 		// max batch size is at least 2, so we'll skip it.

--- a/pkg/sql/backfill/backfill.go
+++ b/pkg/sql/backfill/backfill.go
@@ -288,7 +288,8 @@ func (cb *ColumnBackfiller) RunColumnBackfillChunk(
 	// populated and deleted by the OLTP commands but not otherwise
 	// read or used
 	if err := cb.fetcher.StartScan(
-		ctx, txn, []roachpb.Span{sp}, true /* limitBatches */, chunkSize, traceKV,
+		ctx, txn, []roachpb.Span{sp}, true /* limitBatches */, chunkSize,
+		traceKV, false, /* forceProductionKVBatchSize */
 	); err != nil {
 		log.Errorf(ctx, "scan error: %s", err)
 		return roachpb.Key{}, err
@@ -688,7 +689,8 @@ func (ib *IndexBackfiller) BuildIndexEntriesChunk(
 	// populated and deleted by the OLTP commands but not otherwise
 	// read or used
 	if err := ib.fetcher.StartScan(
-		ctx, txn, []roachpb.Span{sp}, true /* limitBatches */, initBufferSize, traceKV,
+		ctx, txn, []roachpb.Span{sp}, true /* limitBatches */, initBufferSize,
+		traceKV, false, /* forceProductionKVBatchSize */
 	); err != nil {
 		log.Errorf(ctx, "scan error: %s", err)
 		return nil, nil, 0, err

--- a/pkg/sql/colfetcher/cfetcher.go
+++ b/pkg/sql/colfetcher/cfetcher.go
@@ -613,6 +613,7 @@ func (rf *cFetcher) StartScan(
 	limitBatches bool,
 	limitHint int64,
 	traceKV bool,
+	forceProductionKVBatchSize bool,
 ) error {
 	if len(spans) == 0 {
 		return errors.AssertionFailedf("no spans")
@@ -645,6 +646,7 @@ func (rf *cFetcher) StartScan(
 		rf.lockStrength,
 		rf.lockWaitPolicy,
 		nil, /* memMonitor */
+		forceProductionKVBatchSize,
 	)
 	if err != nil {
 		return err

--- a/pkg/sql/colfetcher/colbatch_scan.go
+++ b/pkg/sql/colfetcher/colbatch_scan.go
@@ -71,8 +71,8 @@ func (s *ColBatchScan) Init() {
 	s.init = true
 	limitBatches := !s.parallelize
 	if err := s.rf.StartScan(
-		s.ctx, s.flowCtx.Txn, s.spans,
-		limitBatches, s.limitHint, s.flowCtx.TraceKV,
+		s.ctx, s.flowCtx.Txn, s.spans, limitBatches, s.limitHint, s.flowCtx.TraceKV,
+		s.flowCtx.EvalCtx.TestingKnobs.ForceProductionBatchSizes,
 	); err != nil {
 		colexecerror.InternalError(err)
 	}

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -1326,6 +1326,7 @@ func (t *logicTest) newCluster(serverArgs TestServerArgs) {
 					AssertFuncExprReturnTypes:       true,
 					DisableOptimizerRuleProbability: *disableOptRuleProbability,
 					OptimizerCostPerturbation:       *optimizerCostPerturbation,
+					ForceProductionBatchSizes:       serverArgs.forceProductionBatchSizes,
 				},
 				SQLExecutor: &sql.ExecutorTestingKnobs{
 					DeterministicExplainAnalyze: true,
@@ -1666,8 +1667,11 @@ func getBlocklistIssueNo(blocklistDirective string) (string, int) {
 }
 
 // processConfigs, given a list of configNames, returns the list of
-// corresponding logicTestConfigIdxs.
-func processConfigs(t *testing.T, path string, defaults configSet, configNames []string) configSet {
+// corresponding logicTestConfigIdxs as well as a boolean indicating whether
+// the test works only in non-metamorphic setting.
+func processConfigs(
+	t *testing.T, path string, defaults configSet, configNames []string,
+) (_ configSet, onlyNonMetamorphic bool) {
 	const blocklistChar = '!'
 	// blocklist is a map from a blocked config to a corresponding issue number.
 	// If 0, there is no associated issue.
@@ -1687,15 +1691,11 @@ func processConfigs(t *testing.T, path string, defaults configSet, configNames [
 	}
 
 	if _, ok := blocklist["metamorphic"]; ok && util.IsMetamorphicBuild() {
-		// We have a metamorphic build and the file has !metamorphic
-		// blocklist directive which effectively skips the file, so we
-		// simply return empty configSet.
-		t.Logf("will skip test %s because it cannot run in a metamorphic build; pass TAGS=crdb_test_off to disable metamorphic builds", path)
-		return configSet{}
+		onlyNonMetamorphic = true
 	}
 	if len(blocklist) != 0 && allConfigNamesAreBlocklistDirectives {
 		// No configs specified, this blocklist applies to the default configs.
-		return applyBlocklistToConfigs(defaults, blocklist)
+		return applyBlocklistToConfigs(defaults, blocklist), onlyNonMetamorphic
 	}
 
 	var configs configSet
@@ -1722,7 +1722,7 @@ func processConfigs(t *testing.T, path string, defaults configSet, configNames [
 		}
 	}
 
-	return configs
+	return configs, onlyNonMetamorphic
 }
 
 // readTestFileConfigs reads any LogicTest directive at the beginning of a
@@ -1734,7 +1734,9 @@ func processConfigs(t *testing.T, path string, defaults configSet, configNames [
 //   # LogicTest: default distsql
 //
 // If the file doesn't contain a directive, the default config is returned.
-func readTestFileConfigs(t *testing.T, path string, defaults configSet) configSet {
+func readTestFileConfigs(
+	t *testing.T, path string, defaults configSet,
+) (_ configSet, onlyNonMetamorphic bool) {
 	file, err := os.Open(path)
 	if err != nil {
 		t.Fatal(err)
@@ -1762,7 +1764,7 @@ func readTestFileConfigs(t *testing.T, path string, defaults configSet) configSe
 		}
 	}
 	// No directive found, return the default config.
-	return defaults
+	return defaults, false
 }
 
 type subtestDetails struct {
@@ -3101,6 +3103,9 @@ type TestServerArgs struct {
 	// actually in-memory). If it is unset, then the default limit of 100MB
 	// will be used.
 	tempStorageDiskLimit int64
+	// If set, mutations.MaxBatchSize and row.getKVBatchSize will be overridden
+	// to use the non-test value.
+	forceProductionBatchSizes bool
 }
 
 // RunLogicTest is the main entry point for the logic test. The globs parameter
@@ -3167,6 +3172,9 @@ func RunLogicTestWithDefaultConfig(
 	// Read the configuration directives from all the files and accumulate a list
 	// of paths per config.
 	configPaths := make([][]string, len(logicTestConfigs))
+	// nonMetamorphic mirrors configPaths and indicates whether a particular
+	// config on a particular path can only run in non-metamorphic setting.
+	nonMetamorphic := make([][]bool, len(logicTestConfigs))
 	configDefaults := defaultConfig
 	var configFilter map[string]struct{}
 	if configOverride != "" {
@@ -3186,7 +3194,7 @@ func RunLogicTestWithDefaultConfig(
 		}
 	}
 	for _, path := range paths {
-		configs := readTestFileConfigs(t, path, configDefaults)
+		configs, onlyNonMetamorphic := readTestFileConfigs(t, path, configDefaults)
 		for _, idx := range configs {
 			config := logicTestConfigs[idx]
 			configName := config.name
@@ -3200,6 +3208,7 @@ func RunLogicTestWithDefaultConfig(
 				continue
 			}
 			configPaths[idx] = append(configPaths[idx], path)
+			nonMetamorphic[idx] = append(nonMetamorphic[idx], onlyNonMetamorphic)
 		}
 	}
 
@@ -3216,6 +3225,7 @@ func RunLogicTestWithDefaultConfig(
 	seenPaths := make(map[string]struct{})
 	for idx, cfg := range logicTestConfigs {
 		paths := configPaths[idx]
+		nonMetamorphic := nonMetamorphic[idx]
 		if len(paths) == 0 {
 			continue
 		}
@@ -3230,8 +3240,9 @@ func RunLogicTestWithDefaultConfig(
 			if logicTestsConfigFilter != "" && cfg.name != logicTestsConfigFilter {
 				skip.IgnoreLint(t, "config does not match env var")
 			}
-			for _, path := range paths {
+			for i, path := range paths {
 				path := path // Rebind range variable.
+				onlyNonMetamorphic := nonMetamorphic[i]
 				// Inner test: one per file path.
 				t.Run(filepath.Base(path), func(t *testing.T) {
 					if *rewriteResultsInTestfiles {
@@ -3269,6 +3280,7 @@ func RunLogicTestWithDefaultConfig(
 					if *printErrorSummary {
 						defer lt.printErrorSummary()
 					}
+					serverArgs.forceProductionBatchSizes = onlyNonMetamorphic
 					lt.setup(cfg, serverArgs)
 					lt.runFile(path, cfg)
 

--- a/pkg/sql/mutations/mutations_util.go
+++ b/pkg/sql/mutations/mutations_util.go
@@ -16,19 +16,29 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util"
 )
 
+const productionMaxBatchSize = 10000
+
 var maxBatchSize = defaultMaxBatchSize
 
 var defaultMaxBatchSize = int64(util.ConstantWithMetamorphicTestRange(
-	10000, /* defaultValue */
-	1,     /* min */
-	10000, /* max */
+	productionMaxBatchSize, /* defaultValue */
+	1,                      /* min */
+	productionMaxBatchSize, /* max */
 ))
 
 // MaxBatchSize returns the max number of entries in the KV batch for a
 // mutation operation (delete, insert, update, upsert) - including secondary
 // index updates, FK cascading updates, etc - before the current KV batch is
 // executed and a new batch is started.
-func MaxBatchSize() int {
+//
+// If forceProductionMaxBatchSize is true, then the "production" value will be
+// returned regardless of whether the build is metamorphic or not. This should
+// only be used by tests the output of which differs if maxBatchSize is
+// randomized.
+func MaxBatchSize(forceProductionMaxBatchSize bool) int {
+	if forceProductionMaxBatchSize {
+		return productionMaxBatchSize
+	}
 	return int(atomic.LoadInt64(&maxBatchSize))
 }
 

--- a/pkg/sql/opt/exec/execbuilder/mutation.go
+++ b/pkg/sql/opt/exec/execbuilder/mutation.go
@@ -146,7 +146,7 @@ func (b *Builder) tryBuildFastPathInsert(ins *memo.InsertExpr) (_ execPlan, ok b
 	//     that we send, not a number of rows. We use this as a guideline only,
 	//     and there is no guarantee that we won't produce a bigger batch.)
 	values, ok := ins.Input.(*memo.ValuesExpr)
-	if !ok || values.ChildCount() > mutations.MaxBatchSize() || values.Relational().HasSubquery {
+	if !ok || values.ChildCount() > mutations.MaxBatchSize(false /* forceProductionMaxBatchSize */) || values.Relational().HasSubquery {
 		return execPlan{}, false, nil
 	}
 

--- a/pkg/sql/row/fetcher.go
+++ b/pkg/sql/row/fetcher.go
@@ -569,6 +569,7 @@ func (rf *Fetcher) StartScan(
 	limitBatches bool,
 	limitHint int64,
 	traceKV bool,
+	forceProductionKVBatchSize bool,
 ) error {
 	if len(spans) == 0 {
 		return errors.AssertionFailedf("no spans")
@@ -584,6 +585,7 @@ func (rf *Fetcher) StartScan(
 		rf.lockStrength,
 		rf.lockWaitPolicy,
 		rf.mon,
+		forceProductionKVBatchSize,
 	)
 	if err != nil {
 		return err
@@ -609,6 +611,7 @@ func (rf *Fetcher) StartInconsistentScan(
 	limitBatches bool,
 	limitHint int64,
 	traceKV bool,
+	forceProductionKVBatchSize bool,
 ) error {
 	if len(spans) == 0 {
 		return errors.AssertionFailedf("no spans")
@@ -665,6 +668,7 @@ func (rf *Fetcher) StartInconsistentScan(
 		rf.lockStrength,
 		rf.lockWaitPolicy,
 		rf.mon,
+		forceProductionKVBatchSize,
 	)
 	if err != nil {
 		return err

--- a/pkg/sql/row/fetcher_test.go
+++ b/pkg/sql/row/fetcher_test.go
@@ -173,6 +173,7 @@ func TestNextRowSingle(t *testing.T) {
 				false, /*limitBatches*/
 				0,     /*limitHint*/
 				false, /*traceKV*/
+				false, /*forceProductionKVBatchSize*/
 			); err != nil {
 				t.Fatal(err)
 			}
@@ -293,6 +294,7 @@ func TestNextRowBatchLimiting(t *testing.T) {
 				true,  /*limitBatches*/
 				10,    /*limitHint*/
 				false, /*traceKV*/
+				false, /*forceProductionKVBatchSize*/
 			); err != nil {
 				t.Fatal(err)
 			}
@@ -403,6 +405,7 @@ func TestRowFetcherMemoryLimits(t *testing.T) {
 		false, /*limitBatches*/
 		0,     /*limitHint*/
 		false, /*traceKV*/
+		false, /*forceProductionKVBatchSize*/
 	)
 	assert.Error(t, err)
 	assert.Equal(t, pgerror.GetPGCode(err), pgcode.OutOfMemory)
@@ -490,6 +493,7 @@ INDEX(c)
 		// batch that ends between rows.
 		1,     /*limitHint*/
 		false, /*traceKV*/
+		false, /*forceProductionKVBatchSize*/
 	); err != nil {
 		t.Fatal(err)
 	}
@@ -650,6 +654,7 @@ func TestNextRowSecondaryIndex(t *testing.T) {
 				false, /*limitBatches*/
 				0,     /*limitHint*/
 				false, /*traceKV*/
+				false, /*forceProductionKVBatchSize*/
 			); err != nil {
 				t.Fatal(err)
 			}
@@ -1007,6 +1012,7 @@ func TestNextRowInterleaved(t *testing.T) {
 				false, /*limitBatches*/
 				0,     /*limitHint*/
 				false, /*traceKV*/
+				false, /*forceProductionKVBatchSize*/
 			); err != nil {
 				t.Fatal(err)
 			}

--- a/pkg/sql/row/kv_batch_fetcher.go
+++ b/pkg/sql/row/kv_batch_fetcher.go
@@ -25,21 +25,35 @@ import (
 	"github.com/cockroachdb/errors"
 )
 
-// kvBatchSize is the number of keys we request at a time.
+// getKVBatchSize returns the number of keys we request at a time.
 // On a single node, 1000 was enough to avoid any performance degradation. On
 // multi-node clusters, we want bigger chunks to make up for the higher latency.
+//
+// If forceProductionKVBatchSize is true, then the "production" value will be
+// returned regardless of whether the build is metamorphic or not. This should
+// only be used by tests the output of which differs if defaultKVBatchSize is
+// randomized.
 // TODO(radu): parameters like this should be configurable
-var kvBatchSize int64 = int64(util.ConstantWithMetamorphicTestValue(
-	10000, /* defaultValue */
-	1,     /* metamorphicValue */
+func getKVBatchSize(forceProductionKVBatchSize bool) int64 {
+	if forceProductionKVBatchSize {
+		return productionKVBatchSize
+	}
+	return defaultKVBatchSize
+}
+
+var defaultKVBatchSize = int64(util.ConstantWithMetamorphicTestValue(
+	productionKVBatchSize, /* defaultValue */
+	1,                     /* metamorphicValue */
 ))
+
+const productionKVBatchSize = 10000
 
 // TestingSetKVBatchSize changes the kvBatchFetcher batch size, and returns a function that restores it.
 // This is to be used only in tests - we have no test coverage for arbitrary kv batch sizes at this time.
 func TestingSetKVBatchSize(val int64) func() {
-	oldVal := kvBatchSize
-	kvBatchSize = val
-	return func() { kvBatchSize = oldVal }
+	oldVal := defaultKVBatchSize
+	defaultKVBatchSize = val
+	return func() { defaultKVBatchSize = oldVal }
 }
 
 // sendFunc is the function used to execute a KV batch; normally
@@ -79,6 +93,9 @@ type txnKVFetcher struct {
 	remainingBatches [][]byte
 	mon              *mon.BytesMonitor
 	acc              mon.BoundAccount
+
+	// If set, we will use the production value for kvBatchSize.
+	forceProductionKVBatchSize bool
 }
 
 var _ kvBatchFetcher = &txnKVFetcher{}
@@ -92,6 +109,7 @@ func (f *txnKVFetcher) getBatchSizeForIdx(batchIdx int) int64 {
 	if !f.useBatchLimit {
 		return 0
 	}
+	kvBatchSize := getKVBatchSize(f.forceProductionKVBatchSize)
 	if f.firstBatchLimit == 0 || f.firstBatchLimit >= kvBatchSize {
 		return kvBatchSize
 	}
@@ -192,6 +210,7 @@ func makeKVBatchFetcher(
 	lockStrength descpb.ScanLockingStrength,
 	lockWaitPolicy descpb.ScanLockingWaitPolicy,
 	mon *mon.BytesMonitor,
+	forceProductionKVBatchSize bool,
 ) (txnKVFetcher, error) {
 	sendFn := func(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.BatchResponse, error) {
 		res, err := txn.Send(ctx, ba)
@@ -201,7 +220,8 @@ func makeKVBatchFetcher(
 		return res, nil
 	}
 	return makeKVBatchFetcherWithSendFunc(
-		sendFn, spans, reverse, useBatchLimit, firstBatchLimit, lockStrength, lockWaitPolicy, mon,
+		sendFn, spans, reverse, useBatchLimit, firstBatchLimit, lockStrength,
+		lockWaitPolicy, mon, forceProductionKVBatchSize,
 	)
 }
 
@@ -216,6 +236,7 @@ func makeKVBatchFetcherWithSendFunc(
 	lockStrength descpb.ScanLockingStrength,
 	lockWaitPolicy descpb.ScanLockingWaitPolicy,
 	mon *mon.BytesMonitor,
+	forceProductionKVBatchSize bool,
 ) (txnKVFetcher, error) {
 	if firstBatchLimit < 0 || (!useBatchLimit && firstBatchLimit != 0) {
 		return txnKVFetcher{}, errors.Errorf("invalid batch limit %d (useBatchLimit: %t)",
@@ -261,15 +282,16 @@ func makeKVBatchFetcherWithSendFunc(
 	}
 
 	return txnKVFetcher{
-		sendFn:          sendFn,
-		spans:           copySpans,
-		reverse:         reverse,
-		useBatchLimit:   useBatchLimit,
-		firstBatchLimit: firstBatchLimit,
-		lockStrength:    lockStrength,
-		lockWaitPolicy:  lockWaitPolicy,
-		mon:             mon,
-		acc:             mon.MakeBoundAccount(),
+		sendFn:                     sendFn,
+		spans:                      copySpans,
+		reverse:                    reverse,
+		useBatchLimit:              useBatchLimit,
+		firstBatchLimit:            firstBatchLimit,
+		lockStrength:               lockStrength,
+		lockWaitPolicy:             lockWaitPolicy,
+		mon:                        mon,
+		acc:                        mon.MakeBoundAccount(),
+		forceProductionKVBatchSize: forceProductionKVBatchSize,
 	}, nil
 }
 

--- a/pkg/sql/row/kv_fetcher.go
+++ b/pkg/sql/row/kv_fetcher.go
@@ -54,9 +54,11 @@ func NewKVFetcher(
 	lockStrength descpb.ScanLockingStrength,
 	lockWaitPolicy descpb.ScanLockingWaitPolicy,
 	mon *mon.BytesMonitor,
+	forceProductionKVBatchSize bool,
 ) (*KVFetcher, error) {
 	kvBatchFetcher, err := makeKVBatchFetcher(
-		txn, spans, reverse, useBatchLimit, firstBatchLimit, lockStrength, lockWaitPolicy, mon,
+		txn, spans, reverse, useBatchLimit, firstBatchLimit, lockStrength,
+		lockWaitPolicy, mon, forceProductionKVBatchSize,
 	)
 	return newKVFetcher(&kvBatchFetcher), err
 }

--- a/pkg/sql/rowexec/inverted_joiner.go
+++ b/pkg/sql/rowexec/inverted_joiner.go
@@ -437,7 +437,8 @@ func (ij *invertedJoiner) readInput() (invertedJoinerState, *execinfrapb.Produce
 	log.VEventf(ij.Ctx, 1, "scanning %d spans", len(indexSpans))
 	if err = ij.fetcher.StartScan(
 		ij.Ctx, ij.FlowCtx.Txn, indexSpans, false /* limitBatches */, 0, /* limitHint */
-		ij.FlowCtx.TraceKV); err != nil {
+		ij.FlowCtx.TraceKV, ij.EvalCtx.TestingKnobs.ForceProductionBatchSizes,
+	); err != nil {
 		ij.MoveToDraining(err)
 		return ijStateUnknown, ij.DrainHelper()
 	}

--- a/pkg/sql/rowexec/joinreader.go
+++ b/pkg/sql/rowexec/joinreader.go
@@ -577,7 +577,8 @@ func (jr *joinReader) readInput() (
 	log.VEventf(jr.Ctx, 1, "scanning %d spans", len(spans))
 	if err := jr.fetcher.StartScan(
 		jr.Ctx, jr.FlowCtx.Txn, spans, jr.shouldLimitBatches, 0, /* limitHint */
-		jr.FlowCtx.TraceKV); err != nil {
+		jr.FlowCtx.TraceKV, jr.EvalCtx.TestingKnobs.ForceProductionBatchSizes,
+	); err != nil {
 		jr.MoveToDraining(err)
 		return jrStateUnknown, nil, jr.DrainHelper()
 	}

--- a/pkg/sql/rowexec/rowfetcher.go
+++ b/pkg/sql/rowexec/rowfetcher.go
@@ -32,7 +32,8 @@ import (
 // collector wrapper can be plugged in.
 type rowFetcher interface {
 	StartScan(
-		_ context.Context, _ *kv.Txn, _ roachpb.Spans, limitBatches bool, limitHint int64, traceKV bool,
+		_ context.Context, _ *kv.Txn, _ roachpb.Spans, limitBatches bool,
+		limitHint int64, traceKV bool, forceProductionKVBatchSize bool,
 	) error
 	StartInconsistentScan(
 		_ context.Context,
@@ -43,6 +44,7 @@ type rowFetcher interface {
 		limitBatches bool,
 		limitHint int64,
 		traceKV bool,
+		forceProductionKVBatchSize bool,
 	) error
 
 	NextRow(ctx context.Context) (

--- a/pkg/sql/rowexec/scrub_tablereader.go
+++ b/pkg/sql/rowexec/scrub_tablereader.go
@@ -219,8 +219,8 @@ func (tr *scrubTableReader) Start(ctx context.Context) context.Context {
 	log.VEventf(ctx, 1, "starting")
 
 	if err := tr.fetcher.StartScan(
-		ctx, tr.FlowCtx.Txn, tr.spans,
-		true /* limit batches */, tr.limitHint, tr.FlowCtx.TraceKV,
+		ctx, tr.FlowCtx.Txn, tr.spans, true /* limit batches */, tr.limitHint,
+		tr.FlowCtx.TraceKV, tr.EvalCtx.TestingKnobs.ForceProductionBatchSizes,
 	); err != nil {
 		tr.MoveToDraining(err)
 	}

--- a/pkg/sql/rowexec/stats.go
+++ b/pkg/sql/rowexec/stats.go
@@ -112,9 +112,10 @@ func (c *rowFetcherStatCollector) StartScan(
 	limitBatches bool,
 	limitHint int64,
 	traceKV bool,
+	forceProductionKVBatchSize bool,
 ) error {
 	start := timeutil.Now()
-	err := c.Fetcher.StartScan(ctx, txn, spans, limitBatches, limitHint, traceKV)
+	err := c.Fetcher.StartScan(ctx, txn, spans, limitBatches, limitHint, traceKV, forceProductionKVBatchSize)
 	c.startScanStallTime += timeutil.Since(start)
 	return err
 }
@@ -129,10 +130,11 @@ func (c *rowFetcherStatCollector) StartInconsistentScan(
 	limitBatches bool,
 	limitHint int64,
 	traceKV bool,
+	forceProductionKVBatchSize bool,
 ) error {
 	start := timeutil.Now()
 	err := c.Fetcher.StartInconsistentScan(
-		ctx, db, initialTimestamp, maxTimestampAge, spans, limitBatches, limitHint, traceKV,
+		ctx, db, initialTimestamp, maxTimestampAge, spans, limitBatches, limitHint, traceKV, forceProductionKVBatchSize,
 	)
 	c.startScanStallTime += timeutil.Since(start)
 	return err

--- a/pkg/sql/rowexec/tablereader.go
+++ b/pkg/sql/rowexec/tablereader.go
@@ -188,15 +188,16 @@ func (tr *tableReader) Start(ctx context.Context) context.Context {
 	var err error
 	if tr.maxTimestampAge == 0 {
 		err = tr.fetcher.StartScan(
-			ctx, tr.FlowCtx.Txn, tr.spans,
-			limitBatches, tr.limitHint, tr.FlowCtx.TraceKV,
+			ctx, tr.FlowCtx.Txn, tr.spans, limitBatches, tr.limitHint,
+			tr.FlowCtx.TraceKV,
+			tr.EvalCtx.TestingKnobs.ForceProductionBatchSizes,
 		)
 	} else {
 		initialTS := tr.FlowCtx.Txn.ReadTimestamp()
 		err = tr.fetcher.StartInconsistentScan(
-			ctx, tr.FlowCtx.Cfg.DB, initialTS,
-			tr.maxTimestampAge, tr.spans,
+			ctx, tr.FlowCtx.Cfg.DB, initialTS, tr.maxTimestampAge, tr.spans,
 			limitBatches, tr.limitHint, tr.FlowCtx.TraceKV,
+			tr.EvalCtx.TestingKnobs.ForceProductionBatchSizes,
 		)
 	}
 

--- a/pkg/sql/rowexec/zigzagjoiner.go
+++ b/pkg/sql/rowexec/zigzagjoiner.go
@@ -764,6 +764,7 @@ func (z *zigzagJoiner) nextRow(ctx context.Context, txn *kv.Txn) (rowenc.EncDatu
 			true, /* batch limit */
 			zigzagJoinerBatchSize,
 			z.FlowCtx.TraceKV,
+			z.EvalCtx.TestingKnobs.ForceProductionBatchSizes,
 		)
 		if err != nil {
 			return nil, err
@@ -908,6 +909,7 @@ func (z *zigzagJoiner) maybeFetchInitialRow() error {
 			true, /* batch limit */
 			zigzagJoinerBatchSize,
 			z.FlowCtx.TraceKV,
+			z.EvalCtx.TestingKnobs.ForceProductionBatchSizes,
 		)
 		if err != nil {
 			log.Errorf(z.Ctx, "scan error: %s", err)

--- a/pkg/sql/sem/tree/eval.go
+++ b/pkg/sql/sem/tree/eval.go
@@ -3223,6 +3223,9 @@ type EvalContextTestingKnobs struct {
 	// cost of each expression in the query tree for the purpose of creating
 	// alternate query plans in the optimizer.
 	OptimizerCostPerturbation float64
+	// If set, mutations.MaxBatchSize and row.getKVBatchSize will be overridden
+	// to use the non-test value.
+	ForceProductionBatchSizes bool
 
 	CallbackGenerators map[string]*CallbackValueGenerator
 }

--- a/pkg/sql/tablewriter.go
+++ b/pkg/sql/tablewriter.go
@@ -118,13 +118,19 @@ type tableWriterBase struct {
 	// rows contains the accumulated result rows if rowsNeeded is set on the
 	// corresponding tableWriter.
 	rows *rowcontainer.RowContainer
+	// If set, mutations.MaxBatchSize and row.getKVBatchSize will be overridden
+	// to use the non-test value.
+	forceProductionBatchSizes bool
 }
 
-func (tb *tableWriterBase) init(txn *kv.Txn, tableDesc catalog.TableDescriptor) {
+func (tb *tableWriterBase) init(
+	txn *kv.Txn, tableDesc catalog.TableDescriptor, evalCtx *tree.EvalContext,
+) {
 	tb.txn = txn
 	tb.desc = tableDesc
 	tb.b = txn.NewBatch()
-	tb.maxBatchSize = mutations.MaxBatchSize()
+	tb.forceProductionBatchSizes = evalCtx != nil && evalCtx.TestingKnobs.ForceProductionBatchSizes
+	tb.maxBatchSize = mutations.MaxBatchSize(tb.forceProductionBatchSizes)
 }
 
 // flushAndStartNewBatch shares the common flushAndStartNewBatch() code between

--- a/pkg/sql/tablewriter_delete.go
+++ b/pkg/sql/tablewriter_delete.go
@@ -42,8 +42,8 @@ func (*tableDeleter) desc() string { return "deleter" }
 func (td *tableDeleter) walkExprs(_ func(desc string, index int, expr tree.TypedExpr)) {}
 
 // init is part of the tableWriter interface.
-func (td *tableDeleter) init(_ context.Context, txn *kv.Txn, _ *tree.EvalContext) error {
-	td.tableWriterBase.init(txn, td.tableDesc())
+func (td *tableDeleter) init(_ context.Context, txn *kv.Txn, evalCtx *tree.EvalContext) error {
+	td.tableWriterBase.init(txn, td.tableDesc(), evalCtx)
 	return nil
 }
 
@@ -145,7 +145,9 @@ func (td *tableDeleter) deleteAllRowsScan(
 	); err != nil {
 		return resume, err
 	}
-	if err := rf.StartScan(ctx, td.txn, roachpb.Spans{resume}, true /* limit batches */, 0, traceKV); err != nil {
+	if err := rf.StartScan(
+		ctx, td.txn, roachpb.Spans{resume}, true /* limit batches */, 0, traceKV, td.forceProductionBatchSizes,
+	); err != nil {
 		return resume, err
 	}
 
@@ -276,7 +278,9 @@ func (td *tableDeleter) deleteIndexScan(
 	); err != nil {
 		return resume, err
 	}
-	if err := rf.StartScan(ctx, td.txn, roachpb.Spans{resume}, true /* limit batches */, 0, traceKV); err != nil {
+	if err := rf.StartScan(
+		ctx, td.txn, roachpb.Spans{resume}, true /* limit batches */, 0, traceKV, td.forceProductionBatchSizes,
+	); err != nil {
 		return resume, err
 	}
 

--- a/pkg/sql/tablewriter_insert.go
+++ b/pkg/sql/tablewriter_insert.go
@@ -31,8 +31,8 @@ var _ tableWriter = &tableInserter{}
 func (*tableInserter) desc() string { return "inserter" }
 
 // init is part of the tableWriter interface.
-func (ti *tableInserter) init(_ context.Context, txn *kv.Txn, _ *tree.EvalContext) error {
-	ti.tableWriterBase.init(txn, ti.tableDesc())
+func (ti *tableInserter) init(_ context.Context, txn *kv.Txn, evalCtx *tree.EvalContext) error {
+	ti.tableWriterBase.init(txn, ti.tableDesc(), evalCtx)
 	return nil
 }
 

--- a/pkg/sql/tablewriter_update.go
+++ b/pkg/sql/tablewriter_update.go
@@ -31,8 +31,8 @@ var _ tableWriter = &tableUpdater{}
 func (*tableUpdater) desc() string { return "updater" }
 
 // init is part of the tableWriter interface.
-func (tu *tableUpdater) init(_ context.Context, txn *kv.Txn, _ *tree.EvalContext) error {
-	tu.tableWriterBase.init(txn, tu.tableDesc())
+func (tu *tableUpdater) init(_ context.Context, txn *kv.Txn, evalCtx *tree.EvalContext) error {
+	tu.tableWriterBase.init(txn, tu.tableDesc(), evalCtx)
 	return nil
 }
 

--- a/pkg/sql/tablewriter_upsert_opt.go
+++ b/pkg/sql/tablewriter_upsert_opt.go
@@ -99,7 +99,7 @@ var _ tableWriter = &optTableUpserter{}
 func (tu *optTableUpserter) init(
 	ctx context.Context, txn *kv.Txn, evalCtx *tree.EvalContext,
 ) error {
-	tu.tableWriterBase.init(txn, tu.ri.Helper.TableDesc)
+	tu.tableWriterBase.init(txn, tu.ri.Helper.TableDesc, evalCtx)
 
 	// rowsNeeded, set upon initialization, indicates whether or not we want
 	// rows returned from the operation.

--- a/pkg/sql/upsert_test.go
+++ b/pkg/sql/upsert_test.go
@@ -34,7 +34,7 @@ func TestUpsertFastPath(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
 
-	if mutations.MaxBatchSize() == 1 {
+	if mutations.MaxBatchSize(false /* forceProductionMaxBatchSize */) == 1 {
 		// The fast path requires that the max batch size is at least 2, so
 		// we'll skip the test.
 		skip.UnderMetamorphic(t)

